AX: Merge AcessibilityARIAGridRow into AccessibilityTableRow

hoffmanjoshua opened the pull request:
#### 9e1954b8c83d5f2b8dddddb69a5d4321618d681e
<pre>
AX: Merge AcessibilityARIAGridRow into AccessibilityTableRow
<a href="https://bugs.webkit.org/show_bug.cgi?id=296711">https://bugs.webkit.org/show_bug.cgi?id=296711</a>
<a href="https://rdar.apple.com/157139850">rdar://157139850</a>

Reviewed by Tyler Wilcock.

Move the grid row logic into its parent class, AccessibilityTableRow, as part of
our effort to flatten our class structure.

In the cases where AcessibilityARIAGridRow overrode a method and AccessibilityTableRow
didn&apos;t, we can call the superclass method for non-grid rows.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h: Removed.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityTableColumnInstance const):
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::AccessibilityTableRow):
(WebCore::AccessibilityTableRow::create):
(WebCore::AccessibilityTableRow::parentTable const):
(WebCore::AccessibilityTableRow::disclosedRows):
(WebCore::AccessibilityTableRow::disclosedByRow const):
(WebCore::AccessibilityTableRow::isARIATreeGridRow const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: <a href="https://commits.webkit.org/298065@main">https://commits.webkit.org/298065@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e3d12465848044fdfa18aab65c01c23832c55558

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/114079 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/33831 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/24289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/120247 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/64846 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115968 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/34459 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/42391 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86702 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/forms/constraints/infinite_backtracking.html (failure)") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/41697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/117027 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/27435 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/102461 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/67086 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/26625 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/20589 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63948 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96801 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/20706 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/123475 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/41109 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/30630 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/95538 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/41487 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/98670 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/95321 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24302 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/40455 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/18265 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/37259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40985 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/46489 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/40608 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43908 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/42361 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -106,9 +108,9 @@ AccessibilityTable* AccessibilityTableRow::parentTable() const
         // choose another ancestor table as this row's table.
         if (auto* parentTable = dynamicDowncast<AccessibilityTable>(*parent)) {
-            if (parentTable->isExposable())
+            if (parentTable->isExposable() && (!isARIAGridRow() || parentTable->isAriaTable() || elementName() == ElementName::HTML_tr))
```
`(!isARIAGridRow() || parentTable->isAriaTable() || elementName() == ElementName::HTML_tr)`

Can this be extracted to a named variable? Right now this bundle of conditions is confusing

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -106,9 +108,9 @@ AccessibilityTable* AccessibilityTableRow::parentTable() const
+            if (parentTable->isExposable() && (!isARIAGridRow() || parentTable->isAriaTable() || elementName() == ElementName::HTML_tr))
                 return parentTable;
-            if (parentTable->node())
+            if (!isARIAGridRow() && parentTable->node())
```
Can we add a comment explaining why we have special behavior for grid rows?

hoffmanjoshua replied:
Added in latest commit

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -171,5 +173,81 @@ std::optional<unsigned> AccessibilityTableRow::axRowIndex() const
+    AccessibilityChildrenVector disclosedRows;
+
+    if (!isARIATreeGridRow())
+        return AccessibilityObject::disclosedRows();
```
`AccessibilityChildrenVector disclosedRows;` should go below `if (!isARIATreeGridRow())` since we don't use it if that branch is taken (I know you didn't write this, but maybe we should clean some little things up as part of this transition)

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -171,5 +173,81 @@ std::optional<unsigned> AccessibilityTableRow::axRowIndex() const
+    // Search for the previous row that matches the correct level.
+    int index = rowIndex();
+    auto allRows = parent->rows();
+    int rowCount = allRows.size();
```
I think we can remove this local variable

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -171,5 +173,81 @@ std::optional<unsigned> AccessibilityTableRow::axRowIndex() const
+    if (!parent)
+        return false;
+
+    return parent->isTreeGrid();
```
RefPtr parent = parentTable();
return parent && parent->isTreeGrid();

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.h:
```diff
@@ -67,7 +70,11 @@ class AccessibilityTableRow : public AccessibilityRenderObject {
+    bool isARIATreeGridRow() const final;
+
     unsigned m_rowIndex;
+    bool m_isARIAGridRow { false };
```
Is this fitting into padding created by `unsigned m_rowIndex`?

hoffmanjoshua replied:
Yeah, there were 4 bytes of available padding here so figured we should keep this inside the class.

twilco commented on Source/WebCore/accessibility/AccessibilityTableRow.cpp:
```diff
@@ -171,5 +173,81 @@ std::optional<unsigned> AccessibilityTableRow::axRowIndex() const
+
+    unsigned level = hierarchicalLevel();
+    auto allRows = parent->rows();
+    int rowCount = allRows.size();
```
I don't think this local variable is necessary

hoffmanjoshua committed:
AX: Merge AcessibilityARIAGridRow into AccessibilityTableRow
https://bugs.webkit.org/show_bug.cgi?id=296711
rdar://157139850

Reviewed by Tyler Wilcock.

Move the grid row logic into its parent class, AccessibilityTableRow, as part of
our effort to flatten our class structure.

In the cases where AcessibilityARIAGridRow overrode a method and AccessibilityTableRow
didn't, we can call the superclass method for non-grid rows.

* Source/WebCore/SaferCPPExpectations/UncountedLambdaCapturesCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::createObjectFromRenderer):
(WebCore::AXObjectCache::createFromNode):
* Source/WebCore/accessibility/AccessibilityARIAGridRow.cpp: Removed.
* Source/WebCore/accessibility/AccessibilityARIAGridRow.h: Removed.
* Source/WebCore/accessibility/AccessibilityObject.h:
(WebCore::AccessibilityObject::isAccessibilityTableColumnInstance const):
(WebCore::AccessibilityObject::isAccessibilityARIAGridRowInstance const): Deleted.
* Source/WebCore/accessibility/AccessibilityTableRow.cpp:
(WebCore::AccessibilityTableRow::AccessibilityTableRow):
(WebCore::AccessibilityTableRow::create):
(WebCore::AccessibilityTableRow::parentTable const):
(WebCore::AccessibilityTableRow::disclosedRows):
(WebCore::AccessibilityTableRow::disclosedByRow const):
(WebCore::AccessibilityTableRow::isARIATreeGridRow const):
* Source/WebCore/accessibility/AccessibilityTableRow.h:
* Source/WebCore/accessibility/isolatedtree/AXIsolatedTree.cpp:
(WebCore::createIsolatedObjectData):
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperBase.mm:
* Source/WebCore/accessibility/mac/WebAccessibilityObjectWrapperMac.mm:

Canonical link: https://commits.webkit.org/298065@main
